File: packages/sdk-csharp/test/TinyHumans.Sdk.Tests/IntegrationTest.cs
using System.Text.Json;
using Xunit;
using Xunit.Abstractions;

namespace TinyHumans.Sdk.Tests;

[Trait("Category", "Integration")]
public class IntegrationTest
{
    private readonly ITestOutputHelper _output;

    public IntegrationTest(ITestOutputHelper output)
    {
        _output = output;
    }

    [Fact]
    public async Task FullLifecycle()
    {
        var token = Environment.GetEnvironmentVariable("TINYHUMANS_TOKEN");
        if (string.IsNullOrEmpty(token))
        {
            _output.WriteLine("TINYHUMANS_TOKEN not set, skipping");
            return;
        }

        var ns = $"integration-test-csharp-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";
        using var client = new TinyHumansMemoryClient(token);

        // ── Insert Memory ──
        var nowSeconds = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var insertResp = await client.InsertMemoryAsync(new InsertMemoryParams
        {
            Title = "test-key-1",
            Content = "The capital of France is Paris.",
            Namespace = ns,
            DocumentId = "integration-test-doc-1",
            Metadata = new Dictionary<string, object?> { ["source"] = "integration-test" },
            CreatedAt = nowSeconds,
            UpdatedAt = nowSeconds,
        });
        Assert.True(insertResp.Success, "insert should succeed");
        _output.WriteLine($"Insert: status={insertResp.Status}");

        await Task.Delay(TimeSpan.FromSeconds(2));

        // ── Recall Memory ──
        var recallResp = await client.RecallMemoryAsync(new RecallMemoryParams { Namespace = ns });
        Assert.True(recallResp.Success, "recall should succeed");
        _output.WriteLine($"Recall: cached={recallResp.Cached}");

        // ── Recall Memories (Ebbinghaus) ──
        var memoriesResp = await client.RecallMemoriesAsync(new RecallMemoriesParams { Namespace = ns });
        Assert.True(memoriesResp.Success, "recall memories should succeed");
        _output.WriteLine($"RecallMemories: count={memoriesResp.Memories.Count}");

        // ── Insert Document ──
        try
        {
            var docResp = await client.InsertDocumentAsync(new InsertDocumentParams
            {
                Title = "Test Doc",
                Content = "Document content for integration test",
                Namespace = ns,
            });
            _output.WriteLine($"InsertDocument: {docResp}");
        }
        catch (Exception e) { _output.WriteLine($"InsertDocument: {e.Message}"); }

        // ── Insert Documents Batch ──
        try
        {
            var batchResp = await client.InsertDocumentsBatchAsync(new InsertDocumentsBatchParams
            {
                Documents = new List<InsertDocumentParams>
                {
                    new() { Title = "Batch 1", Content = "Content 1", Namespace = ns },
                    new() { Title = "Batch 2", Content = "Content 2", Namespace = ns },
                },
            });
            _output.WriteLine($"InsertDocumentsBatch: {batchResp}");
        }
        catch (Exception e) { _output.WriteLine($"InsertDocumentsBatch: {e.Message}"); }

        // ── List Documents ──
        try
        {
            var listResp = await client.ListDocumentsAsync(new ListDocumentsParams { Namespace = ns });
            _output.WriteLine($"ListDocuments: {listResp}");
        }
        catch (Exception e) { _output.WriteLine($"ListDocuments: {e.Message}"); }

        // ── Interact Memory ──
        try
        {
            var interactResp = await client.InteractMemoryAsync(new InteractMemoryParams
            {
                Namespace = ns,
                EntityNames = new List<string> { "TEST ENTITY" },
            });
            _output.WriteLine($"InteractMemory: {interactResp}");
        }
        catch (Exception e) { _output.WriteLine($"InteractMemory: {e.Message}"); }

        // ── Record Interactions ──
        try
        {
            var recordResp = await client.RecordInteractionsAsync(new InteractMemoryParams
            {
                Namespace = ns,
                EntityNames = new List<string> { "TEST ENTITY" },
            });
            _output.WriteLine($"RecordInteractions: {recordResp}");
        }
        catch (Exception e) { _output.WriteLine($"RecordInteractions: {e.Message}"); }

        // ── Recall Thoughts ──
        try
        {
            var thoughtsResp = await client.RecallThoughtsAsync(new RecallThoughtsParams { Namespace = ns });
            _output.WriteLine($"RecallThoughts: {thoughtsResp}");
        }
        catch (Exception e) { _output.WriteLine($"RecallThoughts: {e.Message}"); }

        // ── Query Memory Context ──
        try
        {
            var qmcResp = await client.QueryMemoryContextAsync(new QueryMemoryContextParams
            {
                Query = "capital of France",
                Namespace = ns,
            });
            _output.WriteLine($"QueryMemoryContext: {qmcResp}");
        }
        catch (Exception e) { _output.WriteLine($"QueryMemoryContext: {e.Message}"); }

        // ── Get Graph Snapshot ──
        try
        {
            var graphResp = await client.GetGraphSnapshotAsync(new GraphSnapshotParams { Namespace = ns });
            _output.WriteLine($"GetGraphSnapshot: {graphResp}");
        }
        catch (Exception e) { _output.WriteLine($"GetGraphSnapshot: {e.Message}"); }

        // ── Delete Memory ──
        try
        {
            var deleteResp = await client.DeleteMemoryAsync(new DeleteMemoryParams { Namespace = ns });
            _output.WriteLine($"Delete: nodesDeleted={deleteResp.NodesDeleted}");
        }
        catch (Exception e) { _output.WriteLine($"DeleteMemory: {e.Message}"); }
    }
}

File: packages/sdk-dart/test/integration_test.dart
import 'dart:io' show Platform;

import 'package:test/test.dart';

import 'package:tinyhumans_sdk/tinyhumans_sdk.dart';

void main() {
  test('full lifecycle', () async {
    final token = Platform.environment['TINYHUMANS_TOKEN'];
    if (token == null || token.isEmpty) {
      print('TINYHUMANS_TOKEN not set — skipping integration test');
      return;
    }

    final ns =
        'integration-test-dart-${DateTime.now().millisecondsSinceEpoch}';
    final client = TinyHumansMemoryClient(token);

    try {
      // ── Insert Memory ──
      final nowSeconds = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      final insertResp = await client.insertMemory(InsertMemoryParams(
        title: 'test-key-1',
        content: 'The capital of France is Paris.',
        namespace: ns,
        documentId: 'integration-test-doc-1',
        metadata: {'source': 'integration-test'},
        createdAt: nowSeconds,
        updatedAt: nowSeconds,
      ));
      expect(insertResp.success, isTrue, reason: 'insert should succeed');
      print('Insert: status=${insertResp.status}');

      await Future.delayed(const Duration(seconds: 2));

      // ── Recall Memory ──
      final recallResp =
          await client.recallMemory(RecallMemoryParams(namespace: ns));
      expect(recallResp.success, isTrue, reason: 'recall should succeed');
      print('Recall: cached=${recallResp.cached}');

      // ── Recall Memories (Ebbinghaus) ──
      final memoriesResp =
          await client.recallMemories(RecallMemoriesParams(namespace: ns));
      expect(memoriesResp.success, isTrue,
          reason: 'recall memories should succeed');
      print('RecallMemories: count=${memoriesResp.memories.length}');

      // ── Insert Document ──
      try {
        final docResp = await client.insertDocument(InsertDocumentParams(
          title: 'Test Doc',
          content: 'Document content for integration test',
          namespace: ns,
          documentId: 'test-doc-1',
        ));
        print('InsertDocument: $docResp');
      } catch (e) {
        print('InsertDocument: $e');
      }

      // ── Insert Documents Batch ──
      try {
        final batchResp =
            await client.insertDocumentsBatch(InsertDocumentsBatchParams(
          documents: [
            InsertDocumentParams(
                title: 'Batch 1', content: 'Content 1', namespace: ns, documentId: 'batch-doc-1'),
            InsertDocumentParams(
                title: 'Batch 2', content: 'Content 2', namespace: ns, documentId: 'batch-doc-2'),
          ],
        ));
        print('InsertDocumentsBatch: $batchResp');
      } catch (e) {
        print('InsertDocumentsBatch: $e');
      }

      // ── List Documents ──
      try {
        final listResp =
            await client.listDocuments(ListDocumentsParams(namespace: ns));
        print('ListDocuments: $listResp');
      } catch (e) {
        print('ListDocuments: $e');
      }

      // ── Interact Memory ──
      try {
        final interactResp =
            await client.interactMemory(InteractMemoryParams(
          namespace: ns,
          entityNames: ['TEST ENTITY'],
        ));
        print('InteractMemory: $interactResp');
      } catch (e) {
        print('InteractMemory: $e');
      }

      // ── Record Interactions ──
      try {
        final recordResp =
            await client.recordInteractions(InteractMemoryParams(
          namespace: ns,
          entityNames: ['TEST ENTITY'],
        ));
        print('RecordInteractions: $recordResp');
      } catch (e) {
        print('RecordInteractions: $e');
      }

      // ── Recall Thoughts ──
      try {
        final thoughtsResp = await client
            .recallThoughts(RecallThoughtsParams(namespace: ns));
        print('RecallThoughts: $thoughtsResp');
      } catch (e) {
        print('RecallThoughts: $e');
      }

      // ── Query Memory Context ──
      try {
        final qmcResp = await client.queryMemoryContext(
            QueryMemoryContextParams(
                query: 'capital of France', namespace: ns));
        print('QueryMemoryContext: $qmcResp');
      } catch (e) {
        print('QueryMemoryContext: $e');
      }

      // ── Get Graph Snapshot ──
      try {
        final graphResp = await client
            .getGraphSnapshot(GraphSnapshotParams(namespace: ns));
        print('GetGraphSnapshot: $graphResp');
      } catch (e) {
        print('GetGraphSnapshot: $e');
      }

      // ── Delete Memory ──
      try {
        final deleteResp =
            await client.deleteMemory(DeleteMemoryParams(namespace: ns));
        print('Delete: nodesDeleted=${deleteResp.nodesDeleted}');
      } catch (e) {
        print('DeleteMemory: $e');
      }
    } finally {
      client.close();
    }
  });
}
